database/pay.js

```diff
@@ -0,0 +1,110 @@
+          updatePayerBalance(paymentTransaction, paymentDataFromServer, localPaymentInfo),
+          updatePayeeBalance(paymentTransaction, paymentDataFromServer, localPaymentInfo)
+        ])
+      })
+      // commit
+      .then(paymentTransaction.commit)
+      // return the payer's balance
+      .then(() => {
+        res(localPaymentInfo.payerBalance);
+      })
+      .catch(err => {
+        paymentTransaction.rollback;
+        rej(err);
+      })
+    });
+  })
+}
+
+const getPayerBalance = function(paymentTransaction, localPaymentInfo, paymentDataFromServer) {
+  return pg.table('balance')
+  .transacting(paymentTransaction)
+  .select('amount')
+  .where({user_id: paymentDataFromServer.payerId})
+  .then(rows => {
+    localPaymentInfo.payerBalance = parseFloat(rows[0].amount);
+    if(localPaymentInfo.payerBalance < parseFloat(paymentDataFromServer.amount)) {
+      console.error('INSUFFICIENT FUNDS for userId:', paymentDataFromServer.payerId);
+      // return Promise.reject(new Error('Insufficient funds.'));
+      throw new Error('Insufficient funds.');
+    }
+  })
+}
+
+const getPayeeInfo = function(paymentTransaction, localPaymentInfo, paymentDataFromServer) {
+  return pg.table('users')
+  .transacting(paymentTransaction)
+  .select('amount', 'id')
+  .innerJoin('balance', 'users.id', 'balance.user_id')
+  .where({username: paymentDataFromServer.payeeUsername})
+  .then(rows => {
+    // if no user or payer userid === payee userid, throw error
+    if(rows.length === 0 || rows[0].id === parseInt(paymentDataFromServer.payerId)) {
+      console.error('INVALID PAYEE USERNAME:', paymentDataFromServer.payeeUsername);
+      throw new Error('Invalid payee username.');
+    }
+    console.log('select payee balance amount:', rows[0].amount, 'with user id:', rows[0].id);
+    localPaymentInfo.payeeBalance = parseFloat(rows[0].amount);
+    localPaymentInfo.payeeUserId = parseInt(rows[0].id);
+  })
+}
+
+const addTransaction = function(paymentTransaction, txn_id, paymentDataFromServer) {
+  return pg.table('transactions')
+  .transacting(paymentTransaction)
+  .insert({
+    txn_id: parseInt(txn_id[0]),
+    amount: parseFloat(paymentDataFromServer.amount).toFixed(2),
+    note: paymentDataFromServer.note
+  })
+}
+
+const updatePayerBalance = function(paymentTransaction, paymentDataFromServer, localPaymentInfo) {
+  localPaymentInfo.payerBalance -= parseFloat(paymentDataFromServer.amount);
+  return pg.table('balance')
+  .transacting(paymentTransaction)
+  .update({ amount: localPaymentInfo.payerBalance })
+  .where({ user_id: parseInt(paymentDataFromServer.payerId) })
+}
+
+const updatePayeeBalance = function(paymentTransaction, paymentDataFromServer, localPaymentInfo) {
+  localPaymentInfo.payeeBalance += parseFloat(paymentDataFromServer.amount);
+  return pg.table('balance')
+  .transacting(paymentTransaction)
+  .update({ amount: localPaymentInfo.payeeBalance })
+  .where({ user_id: localPaymentInfo.payeeUserId})
+}
+
```

Comment:
This is a lot of code duplicate between updatePayeeBalance and updatePayerBalance. You might consider writing a single function updateUserBalance that can handle both cases.

Reply:
Not quite sure how I'd do this. I can combine into one function, but the updates need to be separate queries. Can discuss.

Reply:
@jGlass314 - yep - whenever you want to refactor, let's strategize together1